### PR TITLE
Event system for logs and web socket communication

### PR DESCRIPTION
**Pull request by douglascamata**
# **THE EVENT SYSTEM!**

**ATTENTION:** This pull request implements the system only, but doesn't use it and the system is subject to changes before being used!

**DISCLAIMER:** The objective of this system is to provide an easy way to log messages (properly using the `logging` module, because we're grown ups) and to send these messages to 3rd party apps through a SocketIO server. Users won't be allowed to add custom code through this system. 

## Example usage inside bot

```
self.event_manager.register_event("location", parameters=['lat', 'lng'])
self.event_manager.emit(
  'location', 
  sender=self, 
  level='info', 
  data={'lat': 1, 'lng':1}
)
```

**NOTE:** parameters presence are validated! If one is missing it will raise `EventMalformedException`!

## Websocket Handler

When enabled and configured, it can listen to any event the bot emits. For example, if bot emits the event `caught_pokemon` with any given JSON-serializable data, any library listening for  `caught_pokemon` will receive this event and the data.

We provide a built-in websocket server that is started with the config variable `websocket_server`. This config variable has url format (host and port together) for the websocket  server to start, like `localhost:4000`.

This local websocket server can be replaced by a remote one and it will be kept simple to be easily portable to any language developers like.

### Future features for websocket server

In the future the websocket handler will be also used as a remote control server for the bot!

## Logging Handler

The logging handler is pretty simple. It tags the log by event sender and log level. The log message right now is `event_name: data_dict`. Like this: 

```
2016-07-29 02:52:08,205 [PokemonGoBot] [ INFO] location: {'lat': 1, 'lng': 1}
```

I have plans to add another parameter to `event_manager.emit` for more beautiful log message.

## Thanks to @halsafar, he started the work on this following my proposal.

### CODE REVIEW

File: pokemongo_bot/__init__.py [hunk 1]
```diff
@@ -39,12 +43,31 @@ def __init__(self, config):
...
         self.cell = None
 
+
     def start(self):
         self._setup_logging()
         self._setup_api()
         self.navigator = SpiralNavigator(self)
         random.seed()
 
+    def _setup_event_system(self):
```
Comment (TheSavior) [inline pokemongo_bot/__init__.py]: Can you put this lower in the file?

File: pokemongo_bot/test/socketio-client.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from socketIO_client import SocketIO, LoggingNamespace, BaseNamespace
+
+
+def on_location(msg):
```
Comment (TheSavior) [inline pokemongo_bot/test/socketio-client.py]: Can we please make this a unit test? It should be run as part of travis. Nobody is ever going to go into this file again to run it and make sure it works. An automated test will ensure that nobody ever breaks it.

File: pokemongo_bot/__init__.py (hunk 1 above)
Comment (douglascamata) [inline pokemongo_bot/__init__.py]: @TheSavior sure, I was putting here to allow easier merge of #1253

Comment (TheSavior) [inline pokemongo_bot/__init__.py]: Why does this belong here and not in the pokecli.py file? If it was in the cli file we'd be able to send an event on startup with the bag information we are currently printing to the console, right?

File: pokemongo_bot/test/socketio-client.py (hunk 2 above)
Comment (douglascamata) [inline pokemongo_bot/test/socketio-client.py]: I can do it (:

File: pokemongo_bot/__init__.py [hunk 3]
```diff
@@ -396,18 +427,14 @@ def _set_starting_position(self):
...
-            try:
-                location_str = self.config.location.encode('utf-8')
-                location = (self._get_pos_by_name(location_str.replace(" ", "")))
-                self.api.set_position(*location)
-                logger.log('')
-                logger.log(u'Location Found: {}'.format(self.config.location))
-                logger.log('GeoPosition: {}'.format(self.position))
-                logger.log('')
-                has_position = True
-            except Exception:
```
Comment (TheSavior) [inline pokemongo_bot/__init__.py]: Why don't we want this catch anymore?

Comment (douglascamata) [inline pokemongo_bot/__init__.py]: It was hiding geolocation errors that shouldn't be hidden, like request denied. But I committed in this PR by mistake. Undo or keep these changes?

**Comment by solderzzc** [comment]
This beautify instruction should be added into wiki. It's very valuable.

File: pokemongo_bot/__init__.py (hunk 1 above)
Comment (douglascamata) [inline pokemongo_bot/__init__.py]: @TheSavior the event system is set up before the api setup (where bag info is being print). We can still use the event system to log it.

File: pokemongo_bot/__init__.py (hunk 3 above)
Comment (TheSavior) [inline pokemongo_bot/__init__.py]: *shrug*

File: pokemongo_bot/socketio_server/app.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import socketio
+import logging
+from eventlet import wsgi
+from flask import Flask, render_template
```
Comment (TheSavior) [inline pokemongo_bot/socketio_server/app.py]: FAILED pylint inspection: E:  4, 0: Unable to import 'flask' (import-error)

Comment (douglascamata) [inline pokemongo_bot/socketio_server/app.py]: 🔔 shaaaaaaaame

**Comment by TheSavior** [comment]
I still want to see that test get hooked up to travis

**Comment by solderzzc** [comment]
Should we merge this? or will be conflict again..

**Comment by solderzzc** [comment]
Have to merge this before conflict and another merge!

**Comment by simonsmh** [comment]
Missing **socketIO_server socketIO_client flask eventlet** in requirements.txt

And after fix it I got this:
  File "pokecli.py", line 40, in <module>
    from pokemongo_bot import PokemonGoBot
  File "/root/PokemonGo-Bot/pokemongo_bot/__init__.py", line
 24, in <module>
    from pokemongo_bot.socketio_server.runner import SocketI
oRunner
  File "/root/PokemonGo-Bot/pokemongo_bot/socketio_server/ru
nner.py", line 6, in <module>
    from app import app, sio
  File "/root/PokemonGo-Bot/pokemongo_bot/socketio_server/ap
p.py", line 8, in <module>
    sio = socketio.Server(async_mode='eventlet', logging=log
ging.NullHandler)
AttributeError: 'module' object has no attribute 'Server'

Any idea?

**Comment by Calcyfer** [comment]
B-E-A-UTIFUL.

**Pull request merged**